core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -3334,20 +3333,14 @@ private void registerOperandSubQueries(
   @Override public void validateLiteral(SqlLiteral literal) {
     switch (literal.getTypeName()) {
     case DECIMAL:
-      // Decimal and long have the same precision (as 64-bit integers), so
-      // the unscaled value of a decimal must fit into a long.
-
-      // REVIEW jvs 4-Aug-2004:  This should probably be calling over to
-      // the available calculator implementations to see what they
-      // support.  For now use ESP instead.
-      //
-      // jhyde 2006/12/21: I think the limits should be baked into the
-      // type system, not dependent on the calculator implementation.
-      BigDecimal bd = literal.getValueAs(BigDecimal.class);
-      BigInteger unscaled = bd.unscaledValue();
-      long longValue = unscaled.longValue();
-      if (!BigInteger.valueOf(longValue).equals(unscaled)) {
-        // overflow
+      // Accept any decimal value that does not exceed the max
+      // precision of the type system.
+      final RelDataTypeSystem typeSystem = getTypeFactory().getTypeSystem();
+      final int maxPrecision = typeSystem.getMaxNumericPrecision();
+      final BigDecimal bd = literal.getValueAs(BigDecimal.class);
+      final BigDecimal noTrailingZeros = bd.stripTrailingZeros();
+      // If we don't strip trailing zeros we may reject values such as 1.000....0.
+      if (noTrailingZeros.precision() > maxPrecision) {
```

Comment:
this code will accept any literal which fits in the specified precision.
I am not sure whether we have to check the scale too.

Reply:
I think we should, I can't think of any reason why we should not, do you have anything in mind?

To this extend, we could use
[BigDecimal::scale()](https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html#scale--) to check compatibility with what the type system supports.

WDYT?

Reply:
I added this check, no tests required changes.

Reply:
Thanks @mihaibudiu, can we add a negative test for it so we make sure it works as intended and it will prevent regressions? After that for me it's good to be merged.

Reply:
I should have done this from the beginning, thank you for reminding me. I added a tiny commit with negative tests.